mysql-test/suite/rocksdb/t/ttl_primary_read_filtering.test

```diff
@@ -0,0 +1,341 @@
+
+select variable_value-@c from information_schema.global_status where variable_name='rocksdb_rows_expired';
+set global rocksdb_compact_cf='default';
+select variable_value-@c from information_schema.global_status where variable_name='rocksdb_rows_expired';
+DROP TABLE t1;
+
+# duplicate PK value attempt to be inserted when old one is expired...
+# in this case, we pretend the expired key was not found and insert into PK
+set global rocksdb_force_flush_memtable_now=1;
+set global rocksdb_compact_cf='default';
+CREATE TABLE t1 (
+  a int PRIMARY KEY
+) ENGINE=rocksdb
+COMMENT='ttl_duration=2;';
+INSERT INTO t1 values (1);
+
+--real_sleep 3
+SELECT * FROM t1;
+
+# this should work, even if old value is not filtered out yet.
+INSERT INTO t1 values (1);
+
+# should show (1) result
+SELECT * FROM t1;
+
+DROP TABLE t1;
+
+# Attempt to update expired value, should filter out
+set global rocksdb_force_flush_memtable_now=1;
+CREATE TABLE t1 (
+  a int PRIMARY KEY
+) ENGINE=rocksdb
+COMMENT='ttl_duration=1;';
+INSERT INTO t1 values (1);
+--real_sleep 2
+
+--sorted_result
+SELECT * FROM t1;
+
+# No error is thrown here, under the hood rnd_next_with_direction is
+# filtering out the record from being seen in the first place.
+UPDATE t1 set a = 1;
+DROP TABLE t1;
+
+##
+## More tests on update behaviour with expired keys.
+##
+set global rocksdb_force_flush_memtable_now=1;
+set global rocksdb_compact_cf='default';
+CREATE TABLE t1 (
+  a int PRIMARY KEY,
+  b int
+) ENGINE=rocksdb
+COMMENT='ttl_duration=4;';
+INSERT INTO t1 values (1,1);
+INSERT INTO t1 values (3,3);
+--real_sleep 5
+
+INSERT INTO t1 values (5,5);
+
+# expired key (1) is still around under the hood, but
+# this time rnd_next_with_direction finds non-expired key (5). So the
+# execution flow in the SQL layer moves onto update_write_row, where it then
+# finds the duplicate key (1). But the duplicate key is expired, so it allows
+# the overwrite.
+UPDATE t1 set a = 1;
+
+--sorted_result
+SELECT * FROM t1;
+
+set global rocksdb_enable_ttl_read_filtering=0;
+# 1,1 should be gone, even with read filtering disabled as it has been
+# overwritten
+--sorted_result
+SELECT * FROM t1;
+set global rocksdb_enable_ttl_read_filtering=1;
+
+# get_row_by_rowid tested here via index_read_map_impl
+UPDATE t1 set a = 999 where a = 1;
+--sorted_result
+SELECT * FROM t1;
+
+UPDATE t1 set a = a - 1;
+--sorted_result
+SELECT * FROM t1;
+
+DROP TABLE t1;
+
+# Ensure no rows can disappear in the middle of long-running transactions
+# Also ensure repeatable-read works as expected
+connect (con1,localhost,root,,);
+connect (con2,localhost,root,,);
+
+set global rocksdb_force_flush_memtable_now = 1;
+set global rocksdb_compact_cf='default';
+CREATE TABLE t1 (
+  a int PRIMARY KEY
+) ENGINE=rocksdb
+COMMENT='ttl_duration=3;';
+
+INSERT INTO t1 values (1);
+INSERT INTO t1 values (3);
+INSERT INTO t1 values (5);
+INSERT INTO t1 values (7);
+
+connection con1;
+--echo # Creating Snapshot (start transaction)
+BEGIN;
```

Comment:
Right, I misunderstood the problem at hand. Now, looking at this snippet, the following should guarantee the desired snapshot contents on testing hosts that insert arbitrary pauses between statements:

```
INSERT INTO t1 values (1);
INSERT INTO t1 values (3);
INSERT INTO t1 values (5);
INSERT INTO t1 values (7);

BEGIN
--let $snapshot_size= `SELECT COUNT(*) FROM t1`
--let $i= 0
while ($snapshot_size != 4)
{
  $i++;
  ROLLBACK;
  .. insert ...
  BEGIN;
  --let $snapshot_size= `SELECT COUNT(*) FROM t1`
}

if ($snapshot_size != 4 && $i == 1000)
{
  --die Your testing host is too slow for reasonable TTL testing
}
```

This should remove any flakiness on reasonable hosts, but cumbersome, still would fail (but at least explicitly) on the worst machines and would have to be applied not only there but even for the code above like 

```
COMMENT='ttl_duration=5;';

INSERT INTO t1 values (1, UNIX_TIMESTAMP());
--real_sleep 6
INSERT INTO t1 values (2, UNIX_TIMESTAMP());
INSERT INTO t1 values (3, UNIX_TIMESTAMP());
```

where 1 might get TTL-filtered too on a slow host.

Perhaps the way to proceed would be to accept the current code as-is (as you report no flakiness failures on your testing), and wait for us to submit patches, if any